features/fout-meldingen.feature

```diff
@@ -0,0 +1,136 @@
+        | naam   | waarde                                         |
+        | title  | Een of meerdere parameters zijn niet correct.  |
+        | status | 400                                            |
+        | detail | Waarde voor parameter 'rsin' is niet opgegeven |
+        En bevat de response de volgende invalidParams
+        | name | reason                |
+        | rsin | geen waarde opgegeven |
+
+    Scenario: Er is geen waarde voor meerdere parameters opgegeven
+        Als '/wozobjecten?rsin=&fields=' wordt aangeroepen
+        Dan bevat de response de volgende kenmerken
+        | naam   | waarde                                                    |
+        | title  | Een of meerdere parameters zijn niet correct.             |
+        | status | 400                                                       |
+        | detail | Waarde voor parameters 'rsin', 'fields' is niet opgegeven |
+        En bevat de response de volgende invalidParams
+        | name   | reason                |
+        | rsin   | geen waarde opgegeven |
+        | fields | geen waarde opgegeven |
+
+Rule: fields parameter bevat geen onbekende kenmerk namen
+
+    Abstract Scenario: Er is één of meerdere onbekende kenmerk namen opgegeven
+        Als '/wozobjecten?rsin=0345100002016017&fields=<fields waarde>' wordt aangeroepen
+        Dan bevat de response de volgende kenmerken
+        | naam   | waarde                                                |
+        | title  | Een of meerdere parameters zijn niet correct.         |
+        | status | 400                                                   |
+        | detail | Parameter 'fields' bevat een (deels) ongeldige waarde |
+        En bevat de response de volgende invalidParams
+        | name   | reason                |
+        | fields | <reason omschrijving> |
+
+        Voorbeelden:
+        | fields waarde              | reason omschrijving                                          |
+        | bestaatniet                | ongeldige waarde: 'bestaatniet' opgegeven                    |
+        | bestaatniet,bestaatookniet | ongeldige waarden: 'bestaatniet', 'bestaatookniet' opgegeven |
+
+Rule: type van parameter waarde is correct
+
+    Scenario: type van waarde van een parameter is niet correct
+        Als '/wozobjecten?postcode=1234AA&huisnummer=A' wordt aangeroepen
+        Dan bevat de response de volgende kenmerken
+        | naam   | waarde                                               |
+        | title  | Een of meerdere parameters zijn niet correct.        |
+        | status | 400                                                  |
+        | detail | Parameter 'huisnummer' is niet van het correcte type |
+        En bevat de response de volgende invalidParams
+        | name       | reason                   |
+        | huisnummer | waarde 'A' is geen getal |
+
+Rule: Een zoek actuele woz objecten aanroep mag slechts één identificatie parameter bevatten
+
```

Comment:
WOZ-objecten

Reply:
Vraag is ook of je hier een foutmelding op moet geven. Dit zijn (mogelijk) elkaar uitsluitende parameters. We geven ook geen warning als je "&huisnummer=12&huisnummer=14" opgeeft. 
Hoe ver ga je in het faciliteren van de consumer ?

Reply:
is het niet duidelijk wat het gedrag is van queryparameters in een REST API? Als ik zoek met postcode=1234AB&huisnummer=10, dan zoek ik allen wozobjecten met postcode==1234AB EN huisnummer==10, dus niet wozobjecten met postcode==1234AB OF huisnummer==10, o.i.d. En als er geen wozobject bestaat met postcode 1234AB EN huisnummer 10, dan geeft de API een lege collectie terug.
Gebruik van combinaties van andere parameters (identificaties) werken gewoon hetzelfde, toch?

Reply:
> Gebruik van combinaties van andere parameters (identificaties) werken gewoon hetzelfde, toch?

Volgens mij wel

Reply:
> 
> 
> is het niet duidelijk wat het gedrag is van queryparameters in een REST API? Als ik zoek met postcode=1234AB&huisnummer=10, dan zoek ik allen wozobjecten met postcode==1234AB EN huisnummer==10, dus niet wozobjecten met postcode==1234AB OF huisnummer==10, o.i.d. En als er geen wozobject bestaat met postcode 1234AB EN huisnummer 10, dan geeft de API een lege collectie terug.
> Gebruik van combinaties van andere parameters (identificaties) werken gewoon hetzelfde, toch?

Wat ik bedoel is dat de interface zegt dat de consumer het volgende kan: zoeken met postcode, zoeken met huisnummer en zoeken met postcode en huisnummer. Wat is het gedrag van zoeken met alleen huisnummer? Alle woz-objecten die voldoen aan het huisnummer, een foutmelding dat dat niet kan, of helemaal niets en geven een lege lijst terug.

Reply:
> 
> 
> Vraag is ook of je hier een foutmelding op moet geven. Dit zijn (mogelijk) elkaar uitsluitende parameters. We geven ook geen warning als je "&huisnummer=12&huisnummer=14" opgeeft.
> Hoe ver ga je in het faciliteren van de consumer ?

"&huisnummer=12&huisnummer=14" is valide input. Zie OAS specs, parameter object, style examples. Dit is een manier om een array in een query parameter op te geven.

Reply:
Wat gebeurt er eigenlijk als je dit doet? @fsamwel krijgen we dan een foutmelding denk je? Je geeft hier twee huisnummers op terwijl je er maar eentje in mag voeren (dat is in ieder geval wel de bedoeling). Is dit valide input?

Reply:
> "&huisnummer=12&huisnummer=14" is valide input. Zie OAS specs, parameter object, style examples. Dit is een manier om een array in een query parameter op te geven.

@MelvLee dit begrijp ik niet, dat kan toch alleen als je huisnummer als array hebt gedefinieerd? Waar/hoe precies staat in de OAS specificaties dat bij onderstaande definitie van huisnummer als type:integer je meerdere huisnummers mag opgeven?
```
     - in: query
          name: huisnummer
          description: Zoek WOZ-objecten met huisnummer"
          required: false
          schema:
            type: integer
            minimum: 1
            maximum: 99999
```

Ik lees namelijk:
> Query parameters can be primitive values, arrays and objects. OpenAPI 3.0 provides several ways to serialize objects and arrays in the query string.

In ons geval specificeren we een huisnummer dus als primitive Value (integer), niet als array of als object. Ik zou daaromverwachten dat "&huisnummer=12&huisnummer=14" een foutmelding moet geven omdat het NIET valide is, omdat huisnummer gedefinieerd is als integer, niet als array.

Reply:
> Wat ik bedoel is dat de interface zegt dat de consumer het volgende kan: zoeken met postcode, zoeken met huisnummer en zoeken met postcode en huisnummer. Wat is het gedrag van zoeken met alleen huisnummer? Alle woz-objecten die voldoen aan het huisnummer, een foutmelding dat dat niet kan, of helemaal niets en geven een lege lijst terug.

Dat is een goede vraag. Standaard (als wij niks zeggen) geeft zoeken met alleen huisnummer alle Woz-objecten met dat huisnummer. Dat is nu alleen niet handig (zijn er waarschijnlijk heel erg veel), omdat we geen paginering hebben opgenomen. Dus moeten we denk ik óf bij huisnummer ook postcode verplichten (en dat opnemen in de specificaties) óf paginering toevoegen. Of we geven, net als in BRP, een foutmelding wanneer er teveel zoekresultaten zijn.